pkg/ingestor/parser/csaf/parser_csaf.go

```diff
@@ -0,0 +1,253 @@
+	if tree.Name == product_ref {
+		purl := tree.Product.IdentificationHelper["purl"]
+		return &purl
+	}
+
+	for _, b := range tree.Branches {
+		purl := findPurl(ctx, b, product_ref)
+		if purl != nil {
+			return purl
+		}
+	}
+
+	return nil
+}
+
+func findProductRef(ctx context.Context, tree csaf.ProductBranch, product_id string) *string {
+	for _, r := range tree.Relationships {
+		if r.FullProductName.ID == product_id {
+			return &r.ProductRef
+		}
+	}
+
+	for _, b := range tree.Branches {
+		pref := findProductRef(ctx, b, product_id)
+		if pref != nil {
+			return pref
+		}
+	}
+	return nil
+}
+
+func findActionStatement(tree *csaf.Vulnerability, product_id string) *string {
+	for _, r := range tree.Remediations {
+		for _, p := range r.ProductIDs {
+			if p == product_id {
+				return &r.Details
+			}
+		}
+	}
+	return nil
+}
+
+func findImpactStatement(tree *csaf.Vulnerability, product_id string) *string {
+	for _, t := range tree.Threats {
+		if t.Category == "impact" {
+			for _, p := range t.ProductIDs {
+				if p == product_id {
+					return &t.Details
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (c *csafParser) findPkgSpec(ctx context.Context, product_id string) (*generated.PkgInputSpec, error) {
+	pref := findProductRef(ctx, c.csaf.ProductTree, product_id)
+	if pref == nil {
+		return nil, fmt.Errorf("unable to locate product reference for id %s", product_id)
+	}
+
+	purl := findPurl(ctx, c.csaf.ProductTree, *pref)
+	if purl == nil {
+		return nil, fmt.Errorf("unable to locate product url for reference %s", *pref)
+	}
+
+	return helpers.PurlToPkg(*purl)
+}
+
+func (c *csafParser) generateVexIngest(ctx context.Context, cve *generated.CVEInputSpec, ghsa *generated.GHSAInputSpec, vuln *csaf.Vulnerability, status string, product_id string) *assembler.VexIngest {
+	logger := logging.FromContext(ctx)
+	vi := &assembler.VexIngest{}
+
+	vd := generated.VexStatementInputSpec{}
+	vd.KnownSince = c.csaf.Document.Tracking.CurrentReleaseDate
+	vd.Origin = c.csaf.Document.Tracking.ID
+	vd.VexJustification = generated.VexJustificationNotProvided
+
+	if vexStatus, ok := vexStatusMap[status]; ok {
+		vd.Status = vexStatus
+	}
+
+	var statement *string
+	if vd.Status == generated.VexStatusNotAffected {
+		statement = findImpactStatement(vuln, product_id)
+	} else {
+		statement = findActionStatement(vuln, product_id)
+	}
+
+	if statement != nil {
+		vd.Statement = *statement
+	}
+
+	for _, flag := range vuln.Flags {
+		found := false
+		for _, pid := range flag.ProductIDs {
+			if pid == product_id {
+				found = true
+			}
+		}
+		if found {
+			if just, ok := justificationsMap[flag.Label]; ok {
+				vd.VexJustification = just
+			}
+		}
+	}
+
+	vi.VexData = &vd
+	vi.CVE = cve
+	vi.GHSA = ghsa
+
+	pkg, err := c.findPkgSpec(ctx, product_id)
+	if err != nil {
+		logger.Warnf("[csaf] unable to locate package for not-affected product %s", product_id)
+		return nil
+	}
+	vi.Pkg = pkg
```

Comment:
Based on the minimum requirements document:

```
[product_id] and [subcomponent_id] MAY be URIs, URLs, hashes, commit IDs, versions, version ranges, dates, date ranges, or any other identification system.
```

In that case we should add support for an artifact (hashes) also.

Reply:
There is a mismatch in the format for ArtifactInputSpec and what's actually defined in CSAF. For instance, hashes in a CSAF document can look like this:

```
"hashes": [
  {
    "filename": "foo.txt",
    "file_hashes": [
       {
         "algorithm": "sha256",
         "value" : "2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae"
       }
    ]
  },
  {
    "filename": "bar.txt",
    "file_hashes": [
       {
         "algorithm": "sha256",
         "value" : "fcde2b2edba56bf408601fb721fe9b5c338d10ee429ea04fae5511b68fbf8fb9"
       }
    ]
  }
]
```

Whereas the ArtifactInputSpec only relates to a single hash for a single artifact.

Reply:
Are these "hashes" describing the artifact that the VEX is generated for? `ArtifactInputSpec` in VEX ingestion case should only be used if instead of the package, a hash is provided to describe the object that the VEX is generated for.

Reply:
Hmm yea i think that works for our ArtifactInput spec for VEX - hope i'm not missing something here, but we should be able to encode the alg + digest. I am fine with having this as is, and as we see handling of different inputs that come about we can update the parser. I suspect there are things that are not purls as well that we may have to handle heuristically.

Only caveat is CertifyVuln doesnt apply to artifacts, because it's meant to be part of the dependents (e.g. if on log4j, the CertifyVuln node points on log4j). I think there is a notion of `VulnReport` that we used to have which we removed. Maybe that's a separate discussion point to have.

Reply:
Let's open an issue for this and we can discuss it there.